Use a random seed in the quorum generator

AndrejMitrovic opened the pull request:
I will see if it's possible to add an additional networking test that tests that the exact expected quorum is generated. We don't expose this outside the node yet.

I've also found a small blocker which I've resolved with a workaround. Currently after generating the genesis block the nodes do not send each other any preimages for the enrollments in that genesis block.

Resolves https://github.com/bpfkorea/agora/issues/990

AndrejMitrovic committed:
tests: add ability to test quorum configs

This will come in handy to be able to test how
the quorum configuration changes when the
validator set changes, or when the preimages
change.

AndrejMitrovic committed:
Use preimages for the random seed

The getRandomSeed() function retrieves all
the active validator's random seeds for
the given block height.

A validator uses this random seed in
the quorum generator.

The onValidatorsChanged() callback has been changed
to take the block height rather than the implicit
latest block height, as during boot we only have the
commitments in the genesis block and *not* the preimages
for the latest block height.

Currently these preimages are only used for the quorum
generator - and the quorums only change when there are
new enrollments (and therefore new commitments).

In the future the quorums will be regularly reshuffled
and we will need to be able to restore the preimage information
when the node boots up (currently missing)

Fixes #990

AndrejMitrovic committed:
Add ability to get the preimage for a height

The getPreimageAt() function retrieves the preimage
for the given block height. If the height is older
than the preimage's current block height, the preimage
will be hashed until the older preimage is retrieved.

AndrejMitrovic committed:
Quorum generator: Add random seed parameter

This parameter adds a certain amount of randomization
in the quorum assignment of each node.

linked0 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -397,6 +397,61 @@ public class ValidatorSet
+
+    ***************************************************************************/
+
+    public PreImageInfo getPreimageAt (const ref Hash enroll_key,
```
This function is kind of duplicate with the `getPreimage` function. So it would be good to merge this function with the original 'getPreimage`.

AndrejMitrovic replied:
I'm not sure I can merge them. The use-cases are different.

Take a look at this call:

https://github.com/bpfkorea/agora/blob/56cd093013a57cd6f37c0fd9f7caaf663d256bd6/source/agora/consensus/ValidatorSet.d#L414

If I add a `height` parameter to `getPreimage`, what do we pass as a height to the `getPreimage` call?

In this case we just want the latest preimage with no condition on the enrollment height or the distance.

linked0 replied:
> If I add a `height` parameter to `getPreimage`, what do we pass as a height to the `getPreimage` call?
> 
> In this case we just want the latest preimage with no condition on the enrollment height or the distance.

There is only one pre-image for the `enroll_key` in the table, so it can be just selected and you can calculate the proper pre-image with several hashing.

AndrejMitrovic replied:
Example 1:
- Validator submitted preimage for block height 20
- We just externalized block height 10
- We want to get the preimage at height 10. So we hash the existing preimage 10 times to get the preimage at height 10.

Example 2:

- The Validator wants to submit the preimage for block height 30
- We should not hash anything here, just validate the new preimage and replace the old one.

There are two separate use-cases here.

linked0 replied:
I thought the function can process with a default parameter for Example 2. That could make the function complex.
Thank you for your explanation.

AndrejMitrovic replied:
It's not really about complexity. It's not possible to combine the two functions together.

In example 1 if the stored preimage is height 9 but we want a preimage for height 10 then the function should return PreImageInfo.init. So we cannot use `Height(ulong.max)` as the default value because it would return the preimage at height 9.

And if we used `Height(0)` as the default value then the function will unnecessarily hash the value when it's used in example 2.

AndrejMitrovic commented:
I'll rewrite the networking test to remove the need to change the genesis block validation workaround. As discussed with Mathias yesterday, we want all the tests to start with the same genesis block.

AndrejMitrovic commented:
> I'll rewrite the networking test to remove the need to change the genesis block validation workaround. As discussed with Mathias yesterday, we want all the tests to start with the same genesis block.

Updated. Now there is a separate test which does the following:

- Starts with 6 enrolled nodes, validator cycle 10
- At the end of the cycle re-enrolls those validators and adds an additional two outsider validators
- Re-enrolls the 8 validators before they expire

We check the quorum configs each time enrollments change, and it shows that the preimages have an effect on the quorum shuffling.

------------

Now the only problem left is that sometimes the test-suite fails with segfaults. It seems to be reproducible on my machine (but the failure is not always in the same place).

AndrejMitrovic commented:
I added this assert into localrest and it triggers:

```diff
diff --git a/source/geod24/LocalRest.d b/source/geod24/LocalRest.d
index e69b6fb..e2d39fa 100644
--- a/source/geod24/LocalRest.d
+++ b/source/geod24/LocalRest.d
@@ -573,6 +573,7 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
         Variant[] await_msgs;

         scope exc = new ExitException();
+        bool has_shutdown;

         void runNode ()
         {
@@ -583,6 +584,9 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
             {
                 static if (is(T == Command))
                 {
+                    if (has_shutdown)
+                        assert(0);
+
                     scheduler.spawn(() => handleCommand(arg, node, control.filter));
                 }
                 else static if (is(T == Response))
@@ -600,6 +604,8 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
                     C.receiveTimeout(self, 10.msecs,
                         (ShutdownCommand!API e)
                         {
+                            has_shutdown = true;
+
                             if (e.callback !is null)
                                 e.callback(node);
                             exc.restart = e.restart;
```

So it seems that there are still events being handled for the node after `shutdown` was called.

AndrejMitrovic commented:
Ah I need to set it to false in the restart loop too. Let's see..

AndrejMitrovic commented:
I'm not sure whether merging https://github.com/bpfkorea/agora/pull/1019 and https://github.com/bpfkorea/agora/pull/1017 had anything to do with it, but the PR is suddenly green.

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -1159,6 +1167,18 @@ private immutable(Block) makeGenesisBlock (in KeyPair[] key_pairs,
+
+        ulong offset = 0;
+        const kp_v = secretKeyToCurveScalar(key_pair.secret);
+
```
```suggestion
```

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -1159,6 +1167,18 @@ private immutable(Block) makeGenesisBlock (in KeyPair[] key_pairs,
+
+        PreImageInfo preimage;
+        preimage.enroll_key = utxo;
+        preimage.distance = cast(ushort)(validator_cycle);
```
```suggestion
        preimage.distance = cast(ushort)validator_cycle;
```

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -683,11 +684,15 @@ private mixin template TestNodeMixin ()
         super(config);
+
+        foreach (const ref preimage; preimages)
+            assert(this.enroll_man.addPreimage(preimage));
```
We already have the `getEnrollmentManager` injection routine. Maybe we can use that instead?

AndrejMitrovic replied:
Hmm no it doesn't work. The issue is the enrollment manager must be instantiated first, then the ledger is instantiated (and it takes the manager in the ctor), then the ledger populates the validator set in the enrollment manager when loading the blocks.

So preimages need to be added after the enrollment manager and the ledger are constructed, but before `onValidatorsChanged` has been called.

AndrejMitrovic commented on source/agora/node/Validator.d:
```diff
@@ -60,6 +60,9 @@ public class Validator : FullNode, API
     /// The current required set of peer keeys to connect to
```
This commit needs a commit message.

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -251,7 +314,7 @@ private const(PublicKey)[] getKeys (size_t count)
 }
 
 /// Create a shorthash from a 64-byte blob for RNG initialization
```
Just a note: the comment was previously wrong (it was 32 bytes), but now it's correct.

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -284,16 +349,18 @@ unittest
 
     Params
         key = the public key of a node
+        rand_seed = the random seed
```
This parameter isn't mentioned in the docs above.

AndrejMitrovic commented on source/agora/node/Validator.d:
```diff
@@ -134,8 +134,9 @@ public class Validator : FullNode, API
             assert(0);
         }
 
+        const rand_seed = hashFull(1);
```
Small note: The first commit does not use the preimages as the seed yet, because that change itself requires additional code. So I split it up into two commits.

AndrejMitrovic commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -397,6 +397,61 @@ public class ValidatorSet
+                contained.
+            height = the desired preimage block height. If it's older than
+                     the preimage's current block height, the preimage will
+                     be hashed until the older preimage is retrieved.
```
```suggestion
                     be hashed until the older preimage is retrieved.
                     Otherwise PreImageInfo.init is returned.
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -471,6 +471,33 @@ public class EnrollmentManager
+
+        Params:
+            keys = the keys to look up
+            height = the desired block height to look up the images in
```
```suggestion
            height = the desired block height to look up the images for
```

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -471,6 +471,33 @@ public class EnrollmentManager
+        foreach (const ref key; keys)
+        {
+            const preimage = this.validator_set.getPreimageAt(key, height);
+            assert(preimage != PreImageInfo.init);  // this should not happen
```
More specifically: If a validator has not revealed their preimage it means it should be removed from the validator set before this function is called. Therefore the `keys` parameter here will not contain that validator's key anymore.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -471,6 +471,33 @@ public class EnrollmentManager
+
+    ***************************************************************************/
+
+    public Hash getRandomSeed (in Hash[] keys, in Height height) @safe nothrow
```
add assert that keys is not empty

AndrejMitrovic replied:
Forgot to write a unittest here. There is a networking unittest, but I think this function should be introduced in its own commit with a unittest.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -471,6 +471,33 @@ public class EnrollmentManager
+        {
+            const preimage = this.validator_set.getPreimageAt(key, height);
+            assert(preimage != PreImageInfo.init);  // this should not happen
+            rand_seed = hashMulti(rand_seed, preimage);
```
Technically this does `hashMulti(Hash.init, preimage)` on the first loop iteration. However I don't think it really matters.

AndrejMitrovic commented on source/agora/test/QuorumPreimage.d:
```diff
@@ -0,0 +1,424 @@
+/*******************************************************************************
+
+    Tests preimage quorum configuration behavior.
```
```suggestion
    Tests preimage quorum generation behavior.
```

AndrejMitrovic commented on source/agora/test/QuorumPreimage.d:
```diff
@@ -0,0 +1,424 @@
+            format("Node %s has quorum config %s. Expected: %s",
+                idx, node.getQuorumConfig(), quorums_1[idx])));
+
+    Transaction makePayTx (in Transaction prev, PublicKey[] addresses, uint index = 0)
```
I guess this should use the new TxBuilder routines. I haven't used them yet personally.

Geod24 replied:
They are still there tho

AndrejMitrovic replied:
I'll remove these.

AndrejMitrovic commented on source/agora/consensus/Quorum.d:
```diff
@@ -284,16 +349,18 @@ unittest
 *******************************************************************************/
 
-private auto getGenerator (PublicKey key)
+private auto getGenerator (PublicKey key, Hash rand_seed)
```
```suggestion
private auto getGenerator (PublicKey key, const ref Hash rand_seed)
```

AndrejMitrovic replied:
Also for the key.

AndrejMitrovic committed:
Implement getRandomSeed()

This EnrollmentManager function generates the
random seed which should be used to feed the
quorum generator with.

It takes the set of enrolled UTXOs and the
desired block height.

If the block height is in the past compared
to the enrolled height + distance of a preimage,
the preimage will be hashed to get the correct
preimage of the desired block height.

getRandomSeed() then concatenates the hash
of all the preimages for the given height,
and returns the hash to be used as a random seed.

Geod24 commented:
Not  a fan of the first commit. Rest looks nice, but I'll probably dig more into it a bit later.

AndrejMitrovic commented:
> Not a fan of the first commit.

Hey that makes two of us!

AndrejMitrovic commented:
Resolved most comments except the comment about using TxBuilder. I'll do this soon.

AndrejMitrovic commented:
I updated the last commit to use the TxBuilder.

It's still quite tricky to make this work because of the requirement to always need 8 transactions to build a single block.

Geod24 commented on source/agora/test/QuorumPreimage.d:
```diff
@@ -0,0 +1,436 @@
+    auto spendable = network.blocks[$ - 1].txs
+        .filter!(tx => tx.type == TxType.Payment)
+        .map!(tx => iota(tx.outputs.length)
+        .map!(idx => TxBuilder(tx, cast(uint)idx)))
```
That indentation is triggering me

AndrejMitrovic replied:
> suggestion pls

Geod24 replied:
The `map` is on `iota`, not on the result of the previous `map`, so +4 indent
```suggestion
            .map!(idx => TxBuilder(tx, cast(uint)idx)))
```

AndrejMitrovic replied:
Oh right. That was a mistake, it's easy to miss.

AndrejMitrovic commented:
> Not a fan of the first commit.

Right, so the thing that is problematic here is the disconnect between making the enrollment outside of the node (to *create* the genesis block) and the node's enrollment manager itself.

With https://github.com/bpfkorea/agora/pull/798/files the node should be able to load the enrollment from the genesis block and then restore its own preimage and send the latest one to the other nodes.

So I think https://github.com/bpfkorea/agora/pull/798 blocks this PR.

AndrejMitrovic commented:
I know what's causing the segfaults. It's a bug in the NetworkManager's usage of hashmaps:

https://github.com/bpfkorea/agora/blob/906709a22dfdd7a045163bc1931a64af6419de04/source/agora/network/NetworkManager.d#L727-L737

See if you can spot the problem!

MichaelKim20 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -471,6 +471,37 @@ public class EnrollmentManager
+
+    public Hash getRandomSeed (in Hash[] keys, in Height height) @safe nothrow
+    {
+        assert(keys.length != 0);
```
Shouldn't we check if the hash values in the `keys` are sorted in order?

AndrejMitrovic replied:
Thanks, I'll add the assert.

AndrejMitrovic commented:
I'll rebase this.

Edit: this will need a bit more work as I ran into #1049

AndrejMitrovic commented:
Rebased and removed the workaround preimage commit.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -686,6 +686,9 @@ public interface TestAPI : ValidatorAPI
 
     public abstract Enrollment createEnrollmentData();
+
+    ///
```
This needs an extra comment as per the other methods.

Geod24 commented on source/agora/consensus/Quorum.d:
```diff
@@ -60,17 +60,20 @@ private enum double THRESHOLD = 0.80;
 *******************************************************************************/
 
 public QuorumConfig buildQuorumConfig ( const ref PublicKey key,
-    in Hash[] utxo_keys, scope UTXOFinder finder ) @safe nothrow
+    in Hash[] utxo_keys, scope UTXOFinder finder, const ref Hash rand_seed ) 
```
`foo (int bar)` not `foo ( int bar )` but that was already the case before this PR

AndrejMitrovic replied:
dfmt bro.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -1256,3 +1291,72 @@ unittest
+    Hash[] utxos;
+
+    // genesisSpendable returns 8 outputs
+    auto pairs = iota(8).map!(idx => WK.Keys[idx]).array;
```
We should really use `repeat` more often :)

Geod24 replied:
Also here you could also write `WK.Keys.byRange().take(8).array`.

AndrejMitrovic replied:
Oh neato. I forgot you added that.

Geod24 reviewed (approved):
Amazing, thanks. The comments don't really need to addressed now, apart from the one on getQuorumConfig
